File: arena/board.py
# a* path finding
# https://medium.com/@nicholas.w.swift/easy-a-star-pathfinding-7e6689c7f7b2

# build board
## square board with ring shape arena
## Spawns 1 & 2
## Walls 0

from typing import List
from hashlib import new

class board:
    def __init__(self, board: list):
        self.board = board
        self.accessible = self.accessible_terrain()
        self.pieces = {} # may be managed else where


    def accessible_terrain(self) -> list:
        accessible = []
        loc = ()
        for y, column in enumerate(self.board):
            for x, cell in enumerate(column):
                if not cell:
                    loc = (x,y)
                    accessible.append(loc)
        return accessible


    def add_piece(self, id: int, data: dict):
        # intake a unit or object
        pass

    def spawn_on_board(self):
        # for piece in pieces add to board
        pass

    def current_loc(self, id: int) -> tuple:
        for y, column in enumerate(self.board):
            for x, cell in enumerate(column):
                if id in cell:
                    return (x,y)


    def move(self, loc: tuple, direction: str, magnitued: int) -> tuple:

        x, y = loc
        new_loc = ()
        
        # n, s, e, w, ne, nw, se, sw
        if direction == "n":
            new_y = y - magnitued
            new_x = x
            new_loc = (new_x, new_y)
        elif direction == "s":
            new_y = y + magnitued
            new_x = x
            new_loc = (new_x, new_y)
        elif direction == "e":
            new_y = y
            new_x = x + magnitued
            new_loc = (new_x, new_y)
        elif direction == "w":
            new_y = y
            new_x = x - magnitued
            new_loc = (new_x, new_y)
        elif direction == "ne":
            new_y = y - magnitued
            new_x = x + magnitued
            new_loc = (new_x, new_y)
        elif direction == "nw":
            new_y = y - magnitued
            new_x = x - magnitued
            new_loc = (new_x, new_y)
        elif direction == "se":
            new_y = y + magnitued
            new_x = x + magnitued
            new_loc = (new_x, new_y)
        elif direction == "sw":
            new_y = y + magnitued
            new_x = x - magnitued
            new_loc = (new_x, new_y)
        else:
            print("unknown path")
            return loc
        return new_loc


    def valid_moves(self, loc: tuple, magnitue: int) -> list:
        valid_moves = []
        valid_move = ()
        directions = ["n", "s", "e", "w", "ne", "nw", "se", "sw"]
        for direction in directions:
            for i in range(magnitue+1):
                move_option = self.move(loc, direction, i)
                if move_option in self.accessible_terrain():
                    valid_move = (move_option, direction)
                    valid_moves.append(valid_move)
        return valid_moves


    def shortest_path(self, start: tuple, end: tuple) -> list:
        
        list_of_moves = []
        
        # TODO: use valid_moves
        # TODO: change to node based a*
        x1, y1 = start
        x2, y2 = end

        key = True
        while key:

                if x1 < x2:
                    x1 = x1 + 1 
                elif x1 > x2:
                    x1 = x1 - 1
                else:
                    x1 = x1

                if y1 < y2:
                    y1 = y1 + 1
                elif  y1 > y2:
                    y1 = y1 - 1
                else:
                    y1 = y1

                cord: tuple = (x1, y1)
                list_of_moves.append(cord)
                if end in list_of_moves:
                    key = False

        return list_of_moves


    def commit_move(self, current_loc: tuple, new_loc: tuple):
        x1, y1 = current_loc
        x2, y2 = new_loc

        # TODO: check for unit type if can move 
        
        # get id at current_loc
        current_loc_ids = self.ids_in_cells(current_loc)
        new_loc_ids = self.ids_in_cells(new_loc)
        # update new_loc with id
        self.board[y2][x2] = list(set().union(current_loc_ids, new_loc_ids))
        # remove id from current_loc
        self.board[y1][x1] = []


    def ids_in_cells(self, current_loc: tuple) -> list:
        x, y = current_loc
        ids : list = self.board[y][x]
        return ids



if __name__ == "__main__":

    map1 = [
        [[0],[0],[0],[0],[0], [], [], [],[0],[0],[0],[0],[0]],
        [[0],[0],[0], [], [], [], [], [], [], [],[0],[0],[0]],
        [[0],[0], [], [], [], [], [], [], [], [], [],[0],[0]],
        [[0], [], [], [], [], [], [], [], [], [1],[], [],[0]],
        [[0], [], [], [], [], [], [], [], [], [], [], [],[0]],
        [ [], [], [], [], [], [], [], [], [], [], [], [], []],
        [ [], [], [], [], [], [], [], [], [], [], [], [], []],
        [ [], [], [], [], [], [], [], [], [], [], [], [], []],
        [[0], [], [], [], [], [], [], [], [], [], [], [],[0]],
        [[0], [], [], [2],[], [], [], [], [], [], [], [],[0]],
        [[0],[0], [], [], [], [], [], [], [], [], [],[0],[0]],
        [[0],[0],[0], [], [], [], [], [], [], [],[0],[0],[0]],
        [[0],[0],[0],[0],[0], [], [], [],[0],[0],[0],[0],[0]]
    ]

    map = board(map1)

    print(map.current_loc(2))

    map.commit_move(current_loc=(3,9), new_loc=(3,10))
    
    print(map.current_loc(2))


    # print(board)
    #print(current_loc(1))
    #commit_move((9, 3), (9, 2))
    #print(board)
    # print(move(current_loc(1),'n', 1))

    #player_1 = valid_moves(current_loc(1), 3)
    #print(player_1)
    #print(current_loc(2))
    #player_2 = valid_moves(current_loc(2))
    #print(player_2)

    # path = shortest_path(start=current_loc(2),end=current_loc(1))
    # print(path)

    

File: go/board.go
package main

import (
	"fmt"
	"testing"
)

type Board struct {
	board      [][][]int
	accessible []Tuple
	pieces     map[int]map[int]Piece
}

type Tuple struct {
	x, y int
}

type Piece struct {
	id   int
	data map[string]interface{}
}

func NewBoard(boardData [][][]int) *Board {
	accessible := make([]Tuple, 0)
	pieces := make(map[int]map[int]Piece)

	for y, row := range boardData {
		for x, cell := range row {
			if len(cell) == 0 {
				accessible = append(accessible, Tuple{x: x, y: y})
			}
		}
	}

	return &Board{
		board:      boardData,
		accessible: accessible,
		pieces:     pieces,
	}
}

func (b *Board) AddPiece(id int, data map[string]interface{}) {
	// Add piece to the board's pieces
}

func (b *Board) CurrentLocation(id int) Tuple {
	for y, row := range b.board {
		for x, cell := range row {
			for _, pieceID := range cell {
				if pieceID == id {
					return Tuple{x: x, y: y}
				}
			}
		}
	}
	return Tuple{}
}

func (b *Board) Move(loc Tuple, direction string, magnitude int) Tuple {
	x, y := loc.x, loc.y
	newX, newY := x, y

	switch direction {
	case "n":
		newY -= magnitude
	case "s":
		newY += magnitude
	case "e":
		newX += magnitude
	case "w":
		newX -= magnitude
	case "ne":
		newX += magnitude
		newY -= magnitude
	case "nw":
		newX -= magnitude
		newY -= magnitude
	case "se":
		newX += magnitude
		newY += magnitude
	case "sw":
		newX -= magnitude
		newY += magnitude
	}

	return Tuple{x: newX, y: newY}
}

func (b *Board) ValidMoves(loc Tuple, magnitude int) []Tuple {
	validMoves := make([]Tuple, 0)
	directions := []string{"n", "s", "e", "w", "ne", "nw", "se", "sw"}

	for _, direction := range directions {
		for i := 0; i <= magnitude; i++ {
			moveOption := b.Move(loc, direction, i)
			if b.isAccessible(moveOption) {
				validMoves = append(validMoves, moveOption)
			}
		}
	}

	return validMoves
}

func (b *Board) isAccessible(loc Tuple) bool {
	for _, tuple := range b.accessible {
		if tuple == loc {
			return true
		}
	}
	return false
}

func (b *Board) CommitMove(currentLoc, newLoc Tuple) {
	// Update the board with the move
}

func (b *Board) IdsInCell(loc Tuple) []int {
	return b.board[loc.y][loc.x]
}

func TestBoard(t *testing.T) {
	boardData := [][][]int{
		{{0}, {0}, {0}, {0}, {0}, {}, {}, {}, {0}, {0}, {0}, {0}, {0}},
		{{0}, {0}, {0}, {}, {}, {}, {}, {}, {}, {}, {0}, {0}, {0}},
		{{0}, {0}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {0}, {0}},
		{{0}, {}, {}, {}, {}, {}, {}, {}, {}, {1}, {}, {}, {0}},
		{{0}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {0}},
		{{}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}},
		{{}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}},
		{{}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}},
		{{0}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {0}},
		{{0}, {}, {}, {2}, {}, {}, {}, {}, {}, {}, {}, {}, {0}},
		{{0}, {0}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {0}, {0}},
		{{0}, {0}, {0}, {}, {}, {}, {}, {}, {}, {}, {0}, {0}, {0}},
		{{0}, {0}, {0}, {0}, {0}, {}, {}, {}, {0}, {0}, {0}, {0}, {0}},
	}

	board := NewBoard(boardData)

	loc2 := board.CurrentLocation(2)
	fmt.Println(loc2)

	board.CommitMove(Tuple{x: 3, y: 9}, Tuple{x: 3, y: 10})

	loc2AfterMove := board.CurrentLocation(2)
	fmt.Println(loc2AfterMove)
}
